pkg/ambassador/tap/conduit/manager.go

```diff
@@ -238,43 +238,44 @@ type streamRetry struct {
 // Open continually tries to open the stream. The function
 // finishes when the stream is successfully opened or when the
 // close function is called.
-// This function is excepted to run as a goroutine.
 func (sr *streamRetry) Open(nspStream *nspAPI.Stream) {
-	sr.mu.Lock()
-	defer sr.mu.Unlock()
-	if sr.getStatus() == ambassadorAPI.StreamStatus_OPEN {
-		return
-	}
-
 	// Set cancelOpen, so the close function could cancel this Open function.
 	sr.ctxMu.Lock()
+	if sr.cancelOpen != nil { // a previous open is still running
+		sr.ctxMu.Unlock()
+		return
+	}
 	var ctx context.Context
 	ctx, sr.cancelOpen = context.WithCancel(context.TODO())
 	sr.ctxMu.Unlock()
+	go func() {
```

Comment:
A golden rule for synchronization primitives is;

**Lock data, _never_ lock code!**

This locks code.